exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -1196,3 +1326,94 @@ func (b *Bitfinex) WsCancelOffer(orderID int64) error {
 func makeRequestInterface(channelName string, data interface{}) []interface{} {
 	return []interface{}{0, channelName, nil, data}
 }
+
+func validateCRC32(book *orderbook.Base, token int) error {
+	// Order ID's need to be sub-sorted in ascending order, this needs to be
+	// done on the main book to ensure that we do not cut price levels out below
+	reOrderByID(book.Bids)
+	reOrderByID(book.Asks)
+
+	// RO precision calculation is based on order ID's and amount values
+	var bids, asks []orderbook.Item
+	for i := 0; i < 25; i++ {
+		if i < len(book.Bids) {
+			bids = append(bids, book.Bids[i])
+		}
+		if i < len(book.Asks) {
+			asks = append(asks, book.Asks[i])
+		}
+	}
+
+	// ensure '-' (negative amount) is passed back to string buffer as
+	// this is needed for calcs - These get swapped if funding rate
+	bidmod := float64(1)
+	if book.IsFundingRate {
+		bidmod = -1
+	}
+
+	askMod := float64(-1)
+	if book.IsFundingRate {
+		askMod = 1
+	}
+
+	var check strings.Builder
+	for i := 0; i < 25; i++ {
+		if i < len(bids) {
+			check.WriteString(strconv.FormatInt(bids[i].ID, 10))
+			check.WriteString(":")
+			check.WriteString(strconv.FormatFloat(bidmod*bids[i].Amount, 'f', -1, 64))
+			check.WriteString(":")
+		}
+
+		if i < len(asks) {
+			check.WriteString(strconv.FormatInt(asks[i].ID, 10))
+			check.WriteString(":")
+			check.WriteString(strconv.FormatFloat(askMod*asks[i].Amount, 'f', -1, 64))
+			check.WriteString(":")
+		}
+	}
+
+	checksumStr := strings.TrimSuffix(check.String(), ":")
+	checksum := crc32.ChecksumIEEE([]byte(checksumStr))
+	if checksum == uint32(token) {
+		return nil
+	}
+	return fmt.Errorf("invalid checksum %d, expected %d",
```

Comment:
Please add asset and pair to aid debugging.
I'm also getting the following:
```
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding UST 4117238931, expected 2891441959
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding USD 3538374318, expected 1742737531
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding BTC 1141843392, expected 2474172479
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding USD 3538374318, expected 2743611013
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding BTC 1833191410, expected 4172016204
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding EDO 1759554129, expected 3308257416
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding BTC 1833191410, expected 4172016204
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding UST 3092897237, expected 1660253372
[ERROR] |  22/12/2020 11:37:18  | routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: invalid checksum for marginfunding USD 3413876450, expected 993424195
```

The errors only occur for margin funding